File: src/backend/GerenciamentoAtivos/GerenciamentoAtivos.Service/Services/SegmentoService.cs
using GerenciamentoAtivos.Domain.Interfaces;
using GerenciamentoAtivos.Domain.Models;
using GerenciamentoAtivos.Service.DTO_s.Segmento;
using GerenciamentoAtivos.Service.Interfaces;

namespace GerenciamentoAtivos.Service.Services
{
    public class SegmentoService(ISegmentoRepository segmentoRepository) : ISegmentoService
    {
        private readonly ISegmentoRepository _segmentoRepository = segmentoRepository;

        public async Task<IEnumerable<SegmentoDto>> Segmentos()
        {
            IEnumerable<SegmentoDto> lstSegmentos = [];

            var result = await _segmentoRepository.GetAllAsync();

            lstSegmentos = result.Select(s => new SegmentoDto
            {
                Id = s.Id,
                Nome = s.Nome,
                Ativo = s.Active,
                Created = s.Created,
                Updated = s.Updated,
            }).ToList();

            return lstSegmentos;
        }

        public async Task<SegmentoDto> SegmentoById(Guid id)
        {
            var segmento = await _segmentoRepository.GetByIdAsync(id);

            SegmentoDto segmentoDto = new()
            {
                Id = segmento.Id,
                Nome = segmento.Nome,
                Ativo = segmento.Active,
                Created = segmento.Created,
                Updated = segmento.Updated,
            };

            return segmentoDto;
        }

        public async Task<SegmentoDto> Create(CriarSegmentoDto data)
        {
            //Validator

            Segmento segmento = new()
            {
                Nome = data.Nome.ToUpper()
            };

            await _segmentoRepository.AddAsync(segmento);
            await _segmentoRepository.SaveChangesAsync();

            return new SegmentoDto
            {
                Id = segmento.Id,
                Nome = segmento.Nome
            };
        }

        public async Task<bool> Update(AtualizarSegmentoDto data)
        {
            var segmento = await _segmentoRepository.GetByIdAsync(data.Id);

            if (segmento == null) return false;

            segmento.Nome = data.Nome.ToUpper();
            segmento.Active = data.Ativo;

            _segmentoRepository.Update(segmento);
            await _segmentoRepository.SaveChangesAsync();

            return true;
        }

        public async Task<bool> DeleteById(Guid id)
        {
            var result = await _segmentoRepository.GetByIdAsync(id);

            if (result == null) return false;

            _segmentoRepository.Delete(result);
            await _segmentoRepository.SaveChangesAsync();
            
            return true;
        }
    }
}

File: src/frontend/WEB/src/Services/SegmentoService.ts
import api from './Api'

//#region Dto's
export interface SegmentoDto {
  id: string
  nome: string
  ativo: boolean
  created: string
  updated?: string
}

export interface CriarSegmentoDto {
  nome: string
}

export interface AtualizarSegmentoDto {
  id: string
  nome: string
  ativo: boolean
}
//#endregion

export const segmentoService = {
  getAll: async (): Promise<SegmentoDto[]> => {
    const response = await api.get<SegmentoDto[]>('/Segmento/Segmentos/')
    return response.data
  },

  getById: async (id: string): Promise<SegmentoDto> => {
    const response = await api.get<SegmentoDto>(`/Segmento/SegmentoByID/${id}`)
    return response.data
  },

  create: async (segmento: CriarSegmentoDto): Promise<CriarSegmentoDto> => {
    const response = await api.post<CriarSegmentoDto>('/Segmento/CriarSegmento/', segmento)
    return response.data
  },

  update: async (segmento: AtualizarSegmentoDto): Promise<void> => {
    await api.patch<SegmentoDto>('/Segmento/AtualizarSegmento/', segmento)
  },

  delete: async (id: string): Promise<void> => {
    await api.delete(`/Segmento/ExcluirSegmento/${id}`)
  },
}
